tests/test_integrals.py

```diff
@@ -688,3 +718,744 @@ def test_biest_interpolators(self):
+        alpha = 0
+        zeta = np.linspace(-np.pi / iota, np.pi / iota, (2 * eq.M_grid) * 4 + 1)
+        grid = get_rtz_grid(
+            eq,
+            rho,
+            alpha,
+            zeta,
+            coordinates="raz",
+            period=(np.inf, 2 * np.pi, np.inf),
+            iota=iota,
+        )
+        data = eq.compute(
+            Bounce1D.required_names
+            + [
+                "cvdrift",
+                "gbdrift",
+                "grad(psi)",
+                "grad(alpha)",
+                "shear",
+                "iota",
+                "psi",
+                "a",
+            ],
+            grid=grid,
+        )
+        np.testing.assert_allclose(data["psi"], psi)
+        np.testing.assert_allclose(data["iota"], iota)
+        assert np.all(data["B^zeta"] > 0)
+        data["Bref"] = 2 * np.abs(psi_boundary) / data["a"] ** 2
+        data["rho"] = rho
+        data["alpha"] = alpha
+        data["zeta"] = zeta
+        data["psi"] = grid.compress(data["psi"])
+        data["iota"] = grid.compress(data["iota"])
+        data["shear"] = grid.compress(data["shear"])
+
+        # Compute analytic approximation.
+        drift_analytic, cvdrift, gbdrift, pitch_inv = TestBounce1D.drift_analytic(data)
+        # Compute numerical result.
+        bounce = Bounce1D(
+            grid.source_grid,
+            data,
+            quad=leggauss(28),  # converges to absolute and relative tolerance of 1e-7
+            Bref=data["Bref"],
+            Lref=data["a"],
+            check=True,
+        )
+        bounce.check_points(*bounce.points(pitch_inv), pitch_inv, plot=False)
+
+        f = Bounce1D.reshape_data(grid.source_grid, cvdrift, gbdrift)
+        drift_numerical_num = bounce.integrate(
+            pitch_inv=pitch_inv,
+            integrand=TestBounce1D.drift_num_integrand,
+            f=f,
+            num_well=1,
+            check=True,
+        )
+        drift_numerical_den = bounce.integrate(
+            pitch_inv=pitch_inv,
+            integrand=TestBounce1D.drift_den_integrand,
+            num_well=1,
+            weight=np.ones(zeta.size),
+            check=True,
+        )
+        drift_numerical = np.squeeze(drift_numerical_num / drift_numerical_den)
+        msg = "There should be one bounce integral per pitch in this example."
+        assert drift_numerical.size == drift_analytic.size, msg
+        np.testing.assert_allclose(
+            drift_numerical, drift_analytic, atol=5e-3, rtol=5e-2
+        )
+
+        TestBounce1D._test_bounce_autodiff(
+            bounce,
+            TestBounce1D.drift_num_integrand,
+            f=f,
+            weight=np.ones(zeta.size),
+        )
+
+        fig, ax = plt.subplots()
+        ax.plot(pitch_inv.squeeze(), drift_analytic)
+        ax.plot(pitch_inv.squeeze(), drift_numerical)
+        return fig
+
+    @staticmethod
+    def _test_bounce_autodiff(bounce, integrand, **kwargs):
```

Comment:
Very nice!
Can you explain to me why the bounce autodiff was failing before and what changes you had to make to fix it?

Reply:
Prior to commit 02afa2c2a698f65d65c0effeaffa0d8d3f0364cb, nan's were leaking into the gradient (https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#nans). The linked commit resolved this issue. If I recall correctly, the main thing was to use a finite value for the `sentinel` flag in the code instead of `np.nan` and also using  the double where trick https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where.